File: Stanford/kargerMinCut.java
package Stanford;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;

public class kargerMinCut {
    private static Integer[] chooseEdge(Map<Integer, List<Integer>> graph) {
        Random rand = new Random();
        Object[] keys = graph.keySet().toArray();
        Integer fromVertex = (Integer) keys[rand.nextInt(keys.length)];
        List<Integer> vertices = graph.get(fromVertex);
        Integer toVertex = vertices.get(rand.nextInt(vertices.size()));
        return new Integer[]{fromVertex, toVertex};
    }


    public static int kargerMinCut(Map<Integer, List<Integer>> graph) {
        while(graph.size() > 2) {
            Integer[] edge = chooseEdge(graph);
            Integer fromVertex = edge[0];
            Integer toVertex = edge[1];
            List<Integer> vertices = graph.get(toVertex);
            for (int i = 0; i < vertices.size(); i++) {
                Integer cur = vertices.get(i);
                graph.get(fromVertex).add(cur);
                graph.get(cur).remove(toVertex);
                graph.get(cur).add(fromVertex);
            }
            graph.get(fromVertex).removeAll(Collections.singleton(fromVertex));
            graph.remove(toVertex);
        }

        return graph.entrySet().iterator().next().getValue().size();
    }


    public static int operate(int n, Map<Integer, List<Integer>> graph) {
        int count = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            Map<Integer, List<Integer>> g = new HashMap<>(graph);
            int minCut = kargerMinCut(g);
            if (minCut < count)
                count = minCut;
        }
        return count;
    }

    public static void main(String[] args) throws FileNotFoundException {
        Scanner in = new Scanner(new File("src/Stanford/kargerMinCut.txt"));
        Map<Integer, List<Integer>> graph = new HashMap<>();
        while (in.hasNextLine()) {
            List<Integer> vertices = new ArrayList<>();
            String line = in.nextLine();
            String[] arr = line.split("\\s+");
            for (int i = 1; i < arr.length; i++) {
                vertices.add(Integer.parseInt(arr[i]));
            }
            graph.put(Integer.parseInt(arr[0]), vertices);
        }
        operate(50, graph);
    }
}

File: Stanford/kargerMinCut.py
import random
import copy

graph = dict()

with open("kargerMinCut.txt", "r") as f:
    try:
        while True:
            vertices = [int(x) for x in f.readline().rstrip().split()]
            if vertices == []:
                break
            graph[vertices[0]] = vertices[1:]
    except(StopIteration):
        pass

def chooseEdge(graph):
    fromVertex = random.choice(list(graph.keys()))
    toVertex = random.choice(graph[fromVertex])
    return fromVertex, toVertex


def kargerMinCut(graph):
    while len(graph) > 2:
        fromVertex, toVertex = chooseEdge(graph)
        graph[fromVertex].extend(graph[toVertex])
        for x in graph[toVertex]:
            graph[x].remove(toVertex)
            graph[x].append(fromVertex)
        while fromVertex in graph[fromVertex]:
            graph[fromVertex].remove(fromVertex)
        del graph[toVertex]

    for key in graph.keys():
        return len(graph[key])


def operate(n):
    count = 100000
    for i in range(n):
        g = copy.deepcopy(graph)
        minCut = kargerMinCut(g)
        if minCut < count:
            count = minCut
    return count
    
if __name__ == '__main__':
    print(operate(50))


